File: go-src/easy/RangeSumOfBst.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

/**
 * @author : 禾几海
 * @date : 2021年04月28日 07:18
 * @difficulty : easy
 * @info submissions: 106596       accepted: 87191      acceptance: 81.80%
 * @see <a href="https://leetcode-cn.com/problems/range-sum-of-bst">938.二叉搜索树的范围和</a>
 * @desc :
 */

//@lc:start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
//var result = 0

func rangeSumBST(root *TreeNode, low int, high int) int {
	//result = 0
	return inOrderTraversal(root, low, high)
	//return result
}

//func inOrderTraversal_1(root *TreeNode, low int, high int) {
//	if root == nil {
//		return
//	}
//	inOrderTraversal_1(root.Left, low, high)
//	if root.Val >= low && root.Val <= high {
//		result += root.Val
//	}
//	inOrderTraversal_1(root.Right, low, high)
//}

func inOrderTraversal(root *TreeNode, low int, high int) int {
	if root == nil {
		return 0
	}
	if root.Val < low {
		return inOrderTraversal(root.Right, low, high)
	}
	if root.Val > high {
		return inOrderTraversal(root.Left, low, high)
	}
	return root.Val + inOrderTraversal(root.Left, low, high) + inOrderTraversal(root.Right, low, high)
}

//@lc:end

func main() {
}

File: src/cn/celess/easy/RangeSumOfBst.java
package cn.celess.easy;

import cn.celess.utils.*;

/**
 * @author : 禾几海
 * @date : 2021年04月27日 11:26
 * @difficulty : easy
 * @info submissions: 104123       accepted: 85076      acceptance: 81.71%
 * @desc :
 * @see <a href="https://leetcode-cn.com/problems/range-sum-of-bst">938.二叉搜索树的范围和</a>
 */
public class RangeSumOfBst {

    //@lc:start

    /**
     * Definition for a binary tree node.
     * public class TreeNode {
     * int val;
     * TreeNode left;
     * TreeNode right;
     * TreeNode() {}
     * TreeNode(int val) { this.val = val; }
     * TreeNode(int val, TreeNode left, TreeNode right) {
     * this.val = val;
     * this.left = left;
     * this.right = right;
     * }
     * }
     */
    class Solution {
        int result = 0;

        public int rangeSumBST(TreeNode root, int low, int high) {
            inOrderTraversal(root, low, high);
            return result;
        }

        void inOrderTraversal(TreeNode root, int low, int high) {
            if (root == null ) return;
            inOrderTraversal(root.left, low, high);
            if ( root.val >= low && root.val <= high) result+= root.val;
            inOrderTraversal(root.right, low, high);
        }
    }
    //@lc:end


    public static void main(String[] args) {
        Solution solution = new RangeSumOfBst().new Solution();
        Assert assertion = Assert.getInstance();
        System.out.println(solution.rangeSumBST(TreeNodeUtil.createTree(10, 5, 15, 3, 7, null, 18), 7, 15));
        System.out.println(); System.out.println(solution.rangeSumBST(TreeNodeUtil.createTree(10, 5, 15, 3, 7, 13, 18, 1, null, 6), 6, 10));
    }
}
